File: digital_death_trip.rb
require 'fileutils'
require 'find'
load 'tools_for_general_use.rb'
load 'tools_for_trove.rb'
load 'tools_for_towns.rb'
load 'tools_for_geojson.rb'

script_directory = File.dirname(__FILE__)

clear_screen()
my_trove_key = read_trove_key()
search_word = 'tragedy'
default_speed = 180
default_output_path_name = File.join(script_directory, 'output_files')
unless File.directory?(default_output_path_name)
   FileUtils.mkdir_p(default_output_path_name)
end
default_town_path_name = File.join(script_directory, 'town_lists')
max_articles_to_read = 3

continue = true
trove_result_file_name = ''

trove_article_id = '35276208'
trove_article_result = fetch_trove_newspaper_article(trove_article_id = trove_article_id, trove_key = my_trove_key)
trove_article_file = write_trove_newspaper_article_to_file(trove_article_result = trove_article_result, trove_article_id = trove_article_id, output_path_name = default_output_path_name)        
puts("HERE!")
exit()

existing_trove_file_list = return_existing_trove_file_list(output_path_name = default_output_path_name)
puts("\nTrove result files already available: #{existing_trove_file_list.size}\n")

say_something("Hello, this is Digital Death Trip.", also_print = true, speed = default_speed)
say_something("Today I am talking to you from a #{operating_system()} operating system.", also_print = true, speed = default_speed)

say_something("\nWould you like to choose a town, or would you like me to make a random selection?", also_print = true, speed = default_speed)
user_input = get_user_input(prompt_text = "Enter town name OR 'random'\nEnter 'random file' or 'rf' for a random existing file (offline)\nEnter 'exit' to cancel")

if (user_input.upcase == 'EXIT') then
   continue = false
elsif ((user_input.upcase == 'RANDOM') or (user_input.upcase == 'R')) then
   search_town = select_random_town_with_user_input(default_speed = default_speed, town_path_name = default_town_path_name)
elsif ((user_input.upcase == 'RANDOM FILE') or (user_input.upcase == 'RF')) then
   puts("\nSelecting random existing Trove file...")
   existing_trove_file_list = return_existing_trove_file_list(output_path_name = default_output_path_name, also_print = true)
   if (existing_trove_file_list.size == 0) then
      puts("Sorry, no existing Trove files found")
      continue = false
   else
      trove_result_file_name = existing_trove_file_list.sample
      puts("Selected file: #{File.basename(trove_result_file_name)}")
      search_town = return_trove_file_search_town(trove_result_file_name)
   end
else
   search_town = user_input
end

puts("Search town: #{search_town}")

if (continue == true and trove_result_file_name == '') then
   say_something("Ok. I will now see if I can find any newspaper references to a #{search_word} in #{search_town}")
   trove_result_file_name = File.join(default_output_path_name, "trove_result_#{search_town}_#{search_word}.csv".gsub(/\s/,"_"))
   trove_api_results = fetch_trove_search_results(search_town, search_word, my_trove_key)
   puts("\nWriting results to file now...")
   result_count = write_trove_search_results(trove_api_results, trove_result_file_name, search_word, search_town)
   puts(result_count)

   if (result_count == 0) then
      continue = false
      say_something("\nSorry, no tragedy results found for #{search_town}")
   end
end

if (continue == true) then
   result_count = count_trove_search_results_from_csv(trove_result_file_name)
   random_article_range = Array(1..result_count)
   say_something("\nI now have #{result_count} results on file.\nWould you like me to read a few headlines, to get a sense of the tragedies in #{search_town}?", also_print = true, speed = default_speed)
   user_input = get_user_input(prompt_text = "Enter 'n' if not interested, \nEnter 'exit' to cancel entirely, \nEnter any other key to hear a few sample headlines...")
   if (user_input.upcase == 'EXIT') then
      continue = false
   elsif (user_input.upcase == 'ALL') then
      read_trove_headlines(input_trove_file = trove_result_file_name, speed = default_speed)
   elsif (user_input.upcase != 'N') then
      # reads random sample of 5
      random_article_numbers = Array.new(5) { rand(1..result_count) }
      random_article_numbers = random_article_numbers.uniq
      read_trove_headlines(input_trove_file = trove_result_file_name, speed = default_speed, article_numbers = random_article_numbers)
   end
end

if (continue == true) then 
   say_something("\nShall I pick a random tragedy from this place?", also_print = true, speed = default_speed)
   say_something("Or let me know if you would like to pick a specific article.", also_print = true, speed = default_speed)  
end

while (continue == true) do      
   user_input = get_user_input(prompt_text = "\nI will default to a random selection. \nPlease enter 'pick' if you would like to pick. \nEnter 'exit' to cancel.")
   if (user_input.upcase == 'EXIT') then
      continue = false
   elsif (user_input.upcase == 'PICK') then 
      say_something("\nWhich article are you interested in?", also_print = true, speed = default_speed)  
      user_input = get_user_input(prompt_text = "\Please enter article number\nEnter 'exit' to cancel")   
      if (user_input.upcase == 'EXIT') then
         continue = false
      else
         article_numbers = return_int_array_from_string(user_input, divider = ",")   
         if (article_numbers == false) then
            continue = false
         else
            say_something("Ok. Let's see.", also_print = true, speed = default_speed)
            article_number = article_numbers[0]
         end
      end
   else
      article_number = random_article_range.sample
      say_something("Ok. Here is my random tragedy from #{search_town}.", also_print = true, speed = default_speed)
      puts(article_number)
   end
   if (continue == true) then  
      read_trove_results_by_array(input_trove_file = trove_result_file_name, article_numbers = [article_number], speed = default_speed)
      say_something("\n...So, that was a tragedy from #{search_town}", also_print = true, speed = default_speed)   
      say_something("\nWould you like me to get a copy of the whole article for you?", also_print = true, speed = default_speed)                  
      user_input = get_user_input(prompt_text = "\nEnter 'n' for a different tragedy\nEnter 'exit' to cancel\nEnter 'y' to find out more")
      if (user_input.upcase == 'Y') then
         trove_article_id = return_record_from_csv_file(input_file = trove_result_file_name, row_number = article_number + 1, column_number = 9)
         trove_article_result = fetch_trove_newspaper_article(trove_article_id = trove_article_id, trove_key = my_trove_key)
         trove_article_file = write_trove_newspaper_article_to_file(trove_article_result = trove_article_result, trove_article_id = trove_article_id, output_path_name = default_output_path_name)        
         if (trove_article_file != false) then
            puts("\nContent written to file: #{trove_article_file}")
         end
         say_something("Good luck!", also_print = true, speed = default_speed)
         continue = false
      elsif (user_input.upcase == 'EXIT') then      
         continue = false
      end
   end
end

say_something("\nThank you, goodbye.", also_print = true, speed = default_speed)

puts("\nWill update map files before exiting...")
current_result = write_geojson_for_all_csv_files(town_path_name = default_town_path_name, output_path_name = default_output_path_name)
if (current_result != false) then
   puts("\nHave written #{current_result} map objects to your output directory.")
   puts("\nYou may find the map files useful.\nYou can open them in QGIS or in Google Maps.")
else
   puts("\nSorry, encountered error with updating map files.")
end

File: digital_death_trip_python/digital_death_trip.py
import os
import random
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, '..')

from tools_for_general_use import *
from tools_for_trove import *
from tools_for_towns import *
from tools_for_language_processing import *

# TODO: move to separate module, include in requirements
from bs4 import BeautifulSoup

clear_screen()

# start variables: will default to tragedy unless command line argument passed
# IDEA: use random word generator
if len(sys.argv) > 1:
    search_word = sys.argv[1]
else:
    search_word = 'tragedy' 

if len(sys.argv) > 2:
    search_town = sys.argv[2]
else:
    search_town = None

# other start variables
default_speed = 180
max_articles_to_read = 3
max_calls = 10
try_say = test_say_something()
continue_script = True
trove_result_file_name = ''

# directory setup
script_directory = os.path.dirname(os.path.abspath(__file__))
parent_directory = return_parent_directory(script_directory)
operating_system = return_operating_system()
trove_key = return_trove_key()
default_output_path_name = os.path.join(script_directory, 'output_files')
os.makedirs(default_output_path_name, exist_ok=True)
default_town_directory = os.path.join(script_directory, 'town_lists')

# check for Trove key
if (trove_key == None):
    prompt_text = "\nNo Trove key found!\nEnter Trove key or 'N' to cancel script...\n"
    user_input = get_user_input(prompt_text=prompt_text)
    if (user_input[0].upper() == 'N'):
        continue_script = False
        sys.exit()
    else:
        trove_key = user_input.strip()
    
# start conversation        
say_something("\nHello, this is Digital Death Trip.", try_say=try_say, also_print=True, speed=default_speed)
say_something(f"Today I am talking to you from a {operating_system} operating system.", try_say=try_say, speed=default_speed)

# choice of town data: block only runs if no town name passed as command line
# TODO: move to function
if (search_town == None):
    say_something("\nWould you like to choose a town, or would you like me to make a random selection?", try_say=try_say, speed=default_speed)
    prompt_options = ['']
    prompt_options.append("Type 'r' or Enter for a random town choice (DEFAULT); OR")
    prompt_options.append("Type town name directly; OR")
    prompt_options.append("Type 'rf' for a random choice from existing output files; OR")
    prompt_options.append("Type 'exit' to cancel")
    prompt_text = "You can:" + "\n-\t".join(prompt_options) + "\n\n"
    user_input = get_user_input(prompt_text)

    # different options for assigning search town, depending on user choice
    if (user_input.upper() == 'EXIT'):
        continue_script = False
    elif ((len(user_input) == 0) or (user_input.upper() == 'R')):
        print("Random town")
        town_list = return_town_list_with_user_input(speed=default_speed,
            try_say=try_say, town_directory=default_town_directory)
        if (len(town_list) > 0):
            search_town = select_random_town_with_user_input(town_list=town_list, try_say=try_say, speed=default_speed)
        else:
            continue_script = False
    elif (user_input.upper() == 'RF'):
        print("\nSelecting random existing Trove file...")
        existing_trove_result_files = return_existing_trove_result_files(output_path_name=default_output_path_name)
        if (len(existing_trove_result_files) == 0):
            print("Sorry, no existing Trove files found")
            continue_script = False
        else:
            trove_result_file_name = random.choice(existing_trove_result_files)
            print(f"Selected file: {os.path.basename(trove_result_file_name)}")
            search_town = return_trove_file_search_town(trove_result_file_name)
            search_word = return_trove_file_search_word(trove_result_file_name)
            result_count = return_trove_file_result_count(trove_result_file_name)
            print(f"\n{result_count} result/s found on file for {search_town}")
            trove_result_df = pandas.read_csv(trove_result_file_name)
    else:
        search_town = user_input.strip().title()

if (continue_script == True):
    print("continuing...")
    print(f"Search town: {search_town}")
    print(f"Search word: {search_word}")

# start search (unless using existing file)
if (continue_script == True and trove_result_file_name == ''):
    say_something(f"Ok. I will now see if I can find any newspaper references to a {search_word} in {search_town}", try_say=try_say, speed=default_speed)
    # TODO: incorporate more file name error catching
    trove_search_url = build_trove_search_url(trove_key=trove_key, search_town=search_town, search_word=search_word)
    trove_search_result = fetch_trove_search_result(trove_key=trove_key, trove_search_url=trove_search_url)
    if (trove_search_result == None):        
        continue_script = False
        say_something(f"\nSorry, no {search_word} results found for {search_town}", try_say=try_say, speed=default_speed)
    else:
        trove_search_result_metadata = parse_trove_result_metadata(trove_search_result=trove_search_result, search_word=search_word, search_town=search_town)
        print(trove_search_result_metadata)
        result_count = trove_search_result_metadata["total"]
        print(f"\n{result_count} total result/s found for {search_town}")
        if (result_count == 0):
            continue_script = False
        else:
            say_something(f"\nI found {result_count} total results.", try_say=try_say, speed=default_speed)
            trove_result_df = parse_trove_result_records_to_df(trove_search_result=trove_search_result, result_metadata=trove_search_result_metadata)
            print(len(trove_result_df))
            print("Before filtering")
            trove_result_df = filter_trove_result_df(trove_result_df)
            print(len(trove_result_df))
            print("After filtering")
            trove_result_file_name = os.path.join(default_output_path_name, f"trove_result_{search_town}_{search_word}.csv")
            print(trove_result_file_name)
            # TODO: give option to create multiple calls or not
            print(f"Writing to {trove_result_file_name}")
            trove_result_df.to_csv(trove_result_file_name, index=False)
            next_url = return_next_url_from_trove_result_metadata(trove_result_metadata=trove_search_result_metadata)
            call_count = 0
            while (next_url != None and call_count < max_calls):
                call_count += 1
                # TODO: move key repeated steps to function
                trove_search_url = next_url
                trove_search_result = fetch_trove_search_result(trove_key=trove_key, trove_search_url=trove_search_url)
                trove_search_result_metadata = parse_trove_result_metadata(trove_search_result=trove_search_result, search_word=search_word, search_town=search_town)
                trove_result_df = parse_trove_result_records_to_df(trove_search_result=trove_search_result, result_metadata=trove_search_result_metadata)
                print(f"Total results from search {search_town} {search_word}: {result_count}")
                print(f"API call number {call_count} of maximum {max_calls}")
                print(len(trove_result_df))
                print("Before filtering")
                trove_result_df = filter_trove_result_df(trove_result_df)
                print(len(trove_result_df))
                print("After filtering")
                print(f"Appending to {trove_result_file_name}")
                trove_result_df.to_csv(trove_result_file_name, mode='a', header=False, index=False)
                next_url = return_next_url_from_trove_result_metadata(trove_result_metadata=trove_search_result_metadata)
                print(f"\n{next_url}")
                # TODO: clear up the clutter above!!
        

# summarise results
if (continue_script == True):
    trove_result_df = pandas.read_csv(trove_result_file_name)
    trove_result_df = filter_trove_result_df(trove_result_df)
    available_result_count = return_trove_file_result_count(trove_result_file_name)
    # TODO: deal with possible zero or near-zero results after filtering
    summary_fields = []
    for field_name in ["year", "trove_article_heading", "heading", "snippet"]:
        if (field_name in trove_result_df):
            summary_fields.append(field_name)
    
    # NOTE: I think this summary is better than the random picker
    print("\n**SUMMARY VIEW**\n")
    print(trove_result_df[summary_fields])

    word_list = return_word_list_from_df(df=trove_result_df, field_list=["heading", "snippet"])
    # print(word_list)
    # TODO: reduce summary work: just the most common words; plus maybe pairs of words
    word_summary_list = return_custom_word_summary_list(input_words_all=word_list)
    
    #print(word_summary_list)
    for summary in word_summary_list:
        say_something(summary, try_say=try_say, speed=default_speed)
    # TODO: reduce this area

    # TODO: read some random headlines
    # TODO: summarise the most common year?


if (continue_script == True): 
    say_something(f"\nShall I pick a random {search_word} from {search_town}?", try_say=try_say, speed=default_speed)
    say_something("Or let me know if you would like to pick a specific article.", try_say=try_say, speed=default_speed)  

selected_article_number = None
while (continue_script == True and selected_article_number == None):
    prompt_text = "\nI will default to a random selection."
    prompt_text += "\nPlease enter 'pick' if you would like to pick. \nEnter 'exit' to cancel.\n"
    user_input = get_user_input(prompt_text = prompt_text)
    
    if (user_input.upper() == 'EXIT'):
        continue_script = False
    elif (user_input.upper() == 'PICK'):
        say_something("\nWhich article are you interested in?", try_say=try_say, speed=default_speed)  
        user_input = get_user_input(prompt_text = "\nPlease enter article number\nEnter 'exit' to cancel.\n")   
        if (user_input.upper() == 'EXIT'):
            continue_script = False
        else:
            try:
                selected_article_number = int(user_input)
                selected_article = trove_result_df.loc[selected_article_number]
                say_something("Ok. Let's see...", try_say=try_say, speed=default_speed)
            except:
                print("\nSorry, error with input, try again...")
                selected_article_number = None        

    else:
        say_something(f"Ok. Let's see. Here is my random {search_word} from {search_town}.", try_say=try_say, speed=default_speed)
        selected_article_number = random.choice(trove_result_df.index)
        selected_article = trove_result_df.loc[selected_article_number]


    if (continue_script == True and selected_article_number != None):
        print("\n***")
        for field_name in ["year", "heading", "snippet"]:
            say_something(field_name.upper(), try_say=try_say, speed=default_speed)
            say_something(selected_article[field_name], try_say=try_say, speed=default_speed)
        print("***\n")

        say_something("\nWould you like to continue with this article? Or would you like to choose another?", try_say=try_say, speed = default_speed)  
        prompt_text = "\nPlease enter 'n' if you would like to pick another article. \nEnter 'exit' to cancel."
        prompt_text += "\nEnter any other key to continue with this article.\n"
        user_input = get_user_input(prompt_text = prompt_text)
        if (user_input.upper() == 'EXIT'):
            continue_script = False
        elif (user_input.upper() == 'N'):
            selected_article_number = None
        else:
            say_something("\nOk, I will continue with this article...", try_say=try_say, speed=default_speed)
            

if (continue_script == True and selected_article_number != None):  
    say_something("\nI am going to fetch the complete article from the Trove API, please wait...", try_say=try_say, speed=default_speed)             
    
    selected_article_json = fetch_trove_newspaper_article(trove_key=trove_key, 
        trove_article_id=selected_article["id"], also_print=True)
    
    if (selected_article_json == None):
        say_something("\nSorry, I encountered an error searching for this article...")
        continue_script = False
    else:
        # TODO: neaten this section / put in function
        file_description = "trove_article_"
        file_description += search_town + "_"
        file_description += str(selected_article["year"]) + "_"
        file_description += str(selected_article["id"])
        
        # 1 write json file
        json_file_name = os.path.join(default_output_path_name, f"{file_description}.json")
        print(selected_article_json)
        with open(json_file_name, 'w') as f:
            json.dump(selected_article_json, f)
        print(f"Full article content written to json file: {html_file_name}")

        # 2 write pdf file
        pdf_file_name = os.path.join(default_output_path_name, f"{file_description}.pdf")
        print(pdf_file_name)
        pdf_url = article_html["pdf"][0]
        print(pdf_url)
        response = requests.get(pdf_url)
        with open(pdf_file_name, 'wb') as f:
            f.write(response.content)
        print(f"PDF written to: {pdf_file_name}")


        # 3 html of article text
        html_file_name = os.path.join(default_output_path_name, f"{file_description}.html")
        print(html_file_name)
        selected_article_html = selected_article_json["articleText"]
        with open(html_file_name, 'w') as f:
            f.write(selected_article_html)


        # 4 read article out loud (optional)
        # prompt_text = "Would you like me to read the full article text?\n"
        # say_something(prompt_text, try_say=try_say, speed=default_speed)
        # prompt_text = "Enter 'y' to hear the full article.\nEnter any other key to skip this and just create output files.\n"
        # user_input = get_user_input(prompt_text=prompt_text)
        # if (user_input.lower() == 'y'): 

        #     soup = BeautifulSoup(selected_article_html)
        #     selected_article_text = soup.get_text()
        #     say_something(selected_article_text, try_say=try_say, speed=default_speed)

say_something("\nThank you, goodbye.", try_say=try_say, speed=default_speed)

# TODO: map potential here
